homeassistant/core.py

```diff
@@ -1230,7 +1219,7 @@ def call(
             return None
 
         try:
-            with timeout(SERVICE_CALL_LIMIT):
+            async with timeout(limit):
                 await asyncio.shield(self._execute_service(handler, service_call))
```

Comment:
I wonder if this shield should be removed and instead the API endpoints that call this method should shield it. Otherwise we can't cancel a blocking script.

Reply:
Hmm, I had forgotten this shield was here. It wasn't a problem before when I used the `script.turn_off` service to effectively cancel script runs. But I still agree that cancelling is the better way, and this shield _will_ prevent that from working. I guess the tests that have been written so far don't check this scenario, and I didn't re-run my ad hoc tests before committing this last set of changes. ☹️ I guess I need to beef up the tests (oh, and I still have to deal with the tests that wait on `asyncio.sleep`.)

What do you think the ramifications might be if we just remove this shield?

Reply:
After thinking about this some more, and doing some experimenting, I realized that the shield was probably originally there to shield the coroutine from the timeout, but unfortunately for now it also shields cancellations from the caller. But that's actually very easily resolved. I replaced this:
```python
        try:
            async with timeout(limit):
                await asyncio.shield(self._execute_service(handler, service_call))
            return True
        except asyncio.TimeoutError:
            return False

```
with this:
```python
        task = self._hass.loop.create_task(self._execute_service(handler, service_call))
        try:
            async with timeout(limit):
                await asyncio.shield(task)
            return True
        except asyncio.TimeoutError:
            return False
        except asyncio.CancelledError:
            task.cancel()
            raise
```
That maintains the previous behavior (including the shield), but now allows cancellations from the caller to also cancel the service task. With this change all the tests passed as-is. (Basically it's just manually creating the task before calling `asyncio.shield` rather than letting `shield` create the task, so now it has a handle to the task it can use to cancel it.)

I also realized (via ad hoc testing) that the limit has more to do with _what is called_ than _who makes the call_. Specifically, no matter who makes the call, if a script is being called, the limit needs to be disabled. So I'm also moving the limit determination from the Script helper to this call by preceding the above with:
```python
        if domain == "script" and service != SERVICE_TURN_OFF:
            limit = None
        else:
            limit = SERVICE_CALL_LIMIT
```
and removing the `limit` parameter I had added to `call` and `async_call`.

The next commit will include these changes as well as the other fix I mentioned. I think they are necessary for this PR.